### DISCUSSION TITLE
Data object in Pytorch geometric

### DISCUSSION THREAD

**Discussion by abhijeetdhakane** [pyg-team/pytorch_geometric#5147] [Q&A]
Is it possible to store two graphs in the PyTorch geometrics Data object?
e.g.

`Data(x=[B, 1100], edge_index=[B, 2, 4400], edge_attr=[B, 4400, 35], y=[B, 2])`
where `B=2` or `B=1`

In the official documentation, it has mentioned that the object should be as follows
`Data(x=[1100], edge_index=[2, 4400], edge_attr=[4400, 35], y=[2])` as per documentation Data object is limited to hold/store single graph object.

> A graph is used to model pairwise relations (edges) between objects (nodes). A single graph in PyG is described by an instance of [torch_geometric.data.Data](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data), which holds the following attributes by default:

(from https://pytorch-geometric.readthedocs.io/en/latest/notes/introduction.html#data-handling-of-graphs)
Can someone clarify, I was trying to create a custom Dataloader.

**Comment by Padarn** [comment]
You can use a `DataLoader` to batch graphs together, for example. 
```
Graph =[graph1, graph2, graph3, ... , graph500]
loader = DataLoader(dataset = Graph , batch_size=10, shuffle=True)
```
When you sample from this loader you will get a single graph that has an extra attribute `batch` indicating which graph each node  feature corresponds to.

Essentially this stores multiple graphs as a single graph that is not connected.

**Comment by rusty1s** [comment]
You can also look up the concept of [`PairData`](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#pairs-of-graphs).

**Comment by abhijeetdhakane** [comment]
I used a `torch.utils.data import Dataset` and stored two graphs separately with `torch_geometric.data import Data` object like `return data_a, data_b`

If I use `torch_geometric.data import DataLoader` with `batch_size=1` I'm getting the right shapes of Data attributes.
e.g. for `data_a`
```
x -> torch.Size([1100])
edge_index -> torch.Size([2, 4400])
edge_attr -> torch.Size([4400, 20])
y -> torch.Size([100])
```
Now, if I changed `batch_size=10` I'll be getting
```
data_a
x -> torch.Size([11000])
edge_index -> torch.Size([2, 44000])
edge_attr -> torch.Size([44000, 20])
y -> torch.Size([1000])
```
I should get something like
```
data_a
x -> torch.Size([10,1100])
edge_index -> torch.Size([10, 2, 4400])
edge_attr -> torch.Size([10, 4400, 20])
y -> torch.Size([10,100])
```

Any advice?

**Reply by Padarn** [reply]
This is not how the batching works in the `torch_geometric` Dataloader, instead there is a new attribute of the batch `index` that in your case will look like
```
index = [0,0,0,....,1,1,1,...., 9,9,9]
```
where the value indicates which node that batch is from.

If you really want to do it this way you could define your own `collate_fn` something like

```
def collate_fn(datas: List[Data]):
    return Data(edge_index=torch.cat([data.edge_index for data in data]))

loader = torch.utils.data.DataLoader([graph1, graph2,...], collate_fn=collate_fn)
```
but `torch_geometric` doesn't have much support for working with edge_index of the shape you're suggesting.

**Reply by rusty1s** [reply]
Also take a look at the [mini-batching tutorial](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html).

**Reply by abhijeetdhakane** [reply]
Thanks @rusty1s and @Padarn 

I used `__inc__` to increment one of the data attribute which isn't taken care by `pytorch_dataloader` by default. Other data attributes `x, edge_index, edge_attr` taken care by `pytorch_dataloader` and get incremented as increased `batch_size`. 

However, my `batch_size` value isn't accurate. i.e. `batch` attribute. 

For
`train_loader = DataLoader(Data_gen, batch_size=2, shuffle=True)`

output should be 
`Batch(x=[2200], edge_index=[2, 8800], edge_attr=[8800, 20], y=[200], batch=[2], ptr=[3])`

instead I am getting
`Batch(x=[2200], edge_index=[2, 8800], edge_attr=[8800, 20], y=[200], batch=[2200], ptr=[3])`

while `num_nodes = 1100`, looks like it just multiplied by `batch * num_nodes`

**Reply by abhijeetdhakane** [reply]
Also checked `batch` attribute which has `tensor([0, 0, 0,  ..., 1, 1, 1])`, 

As @Padarn mentioned values pointing out the `nodes` belongs to particular `batch`, but still I need to input `batch_size` value i.e. `2` here, to another function.

**Reply by rusty1s** [reply]
This looks correct to me. The `batch` attribute is a vector of shape `[num_nodes]` and maps each node to its example in the mini-batch. You can get the `batch_size` by running `data.batch_size`.
